### ISSUE TITLE
Template filenames with metacharacters throw URI::InvalidURIError

### ISSUE THREAD

**Issue by kbruner** [chef/chef-zero#91]
We have a cookbook around our ldap installation, and some of those files have some URI metacharacters.  This throws an exception when they're used against chef-zero (run in test-kitchen, but looking at the source, test-kitchen is not formulating the URIs):

```
      URI::InvalidURIError
           --------------------
           bad URI(is not URI?): http://localhost:8889/file_store/repo/cookbooks/local-ldap/templates/default/slapd.d/cn=config/olcDatabase={0}config.ldif


       Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/local-ldap/recipes/default.rb


            93:   template "/etc/ldap/slapd.d/cn=config/#{f}.ldif" do
            94:     source "slapd.d/cn=config/#{f}.ldif"
            95:     owner node['ldap']['user']
            96:     group node['ldap']['user']

        97:     mode 00640
            98:     variables :olc_root_pw => vault_info[:olc_root_pw]
            99:   end
           100: end
```

The file in question definitely exists in the template directory and is getting loaded into /tmp/kitchen/cache/.  It also works fine on against a chef-server.

**Comment by DanielLaberge** [comment]
Is there any workaround or are we forced to never use spaces in our file names?

**Comment by kbruner** [comment]
Apologies, hadn't had time to look at it.  I'll try to push a fix this weekend.

**Comment by ndemeshchenko** [comment]
does anybody have any solution? this still seems not resolved

**Comment by byplayer** [comment]
I faced same problem.

I wrote a patch to escape URI meta character.
It works for my environment.

https://github.com/chef/chef-zero/pull/204

**Comment by jfortier** [comment]
Ran into this issue as well still a year and half later. Not sure why this hasn't been merged yet.

👍

**Comment by jonarrien** [comment]
I'm having the same issue with whitespaces, but I ensure I've been working with whitespaces until now. 
Looks something got corrupted in my system after a computer crash. 

ChefDK => 0.14.25
Client    => 12.10.24
Berks    => 4.3.3
Kitchen => 1.8.0

Even I reinstall ChefDK, I'm still having same issue. Any help will be appreciated.

**Comment by victorhdamian** [comment]
URI::InvalidURIError: bad URI(is not URI?): chefzero://localhost:8889/file_store/repo/cookbooks/sensu_xxx/files/default/plugins/process/#check_procmon_registered.rb#

/cc @cheeseplus

**Comment by victorhdamian** [comment]
@jkeiser this is the bug we worked on

**Comment by wolftrouble** [comment]
Sadly, even the newest gem doesn't work. I presume because whatever criteria the URI module is using to determine that a URL is valid goes above and beyond whether it's simply been escaped. What's funny is I found this bug by searching for the error only to realize that my predecessor at Ooyala filed it... For literally the exact same piece of code @kbruner (Hi Karen!)  filed this bug over, and she eventually was forced to use chef-solo because of.

I verified the gem is the news version and that it has the patch submitted in https://github.com/chef/chef-zero/pull/204/commits/5628767a865ac9bf882c2fa5f4441ba66579a747. Error still occurs with at least this file:

           Synchronizing Cookbooks:

         ================================================================================
         Error Syncing Cookbooks:
         ================================================================================

         Unexpected Error:
         -----------------
         URI::InvalidURIError: bad URI(is not URI?): chefzero://localhost:8889/file_store/repo/cookbooks/ooyala-ldap/templates/default/slapd.d/cn=config/olcDatabase={-1}frontend.ldif

**Comment by thommay** [comment]
This should be resolved by #204
